Rust/src/lib.rs

```diff
@@ -1,111 +1,147 @@
-                              number_of_coconuts: 1,
-                              voltage: 0.0,
-                              nailed: false };
-        assert_eq!(parrot.speed().unwrap(), 3.0);
+    // parrot_test macro to run tests for types that implement the Parrot trait
+    // syntax: ([type that implements Parrot, Vec<ParrotTest>])
+    macro_rules! parrot_test {
+        ($type: ty, $tests:expr) => {{
+            for test in $tests {
+                let parrot = <$type>::new(test.config);
+                println!("{}", test.description);
+                assert_eq!(parrot.speed().unwrap(), test.expected);
+            }
+        }};
     }
 
     #[test]
-    fn african_parrot_speed_with_two_coconut() {
-        let parrot = Parrot { parrot_type: "african_parrot",
-                              number_of_coconuts: 2,
-                              voltage: 0.0,
-                              nailed: false };
-        assert_eq!(parrot.speed().unwrap(), 0.0);
-    }
+    fn test_european_parrot_speed() {
+        let tests = vec![ParrotTest {
+            expected: 12.0,
+            config: ParrotConfig::new(0, 0.0, false),
+            description: "Test european parrot speed",
+        }];
 
-    #[test]
-    fn african_parrot_speed_with_no_coconut() {
-        let parrot = Parrot { parrot_type: "african_parrot",
-                              number_of_coconuts: 0,
-                              voltage: 0.0,
-                              nailed: false };
-        assert_eq!(parrot.speed().unwrap(), 12.0);
-    }
-    #[test]
-    fn nailed_norwegian_blue_parrot() {
-        let parrot = Parrot { parrot_type: "norwegian_blue_parrot",
-                              number_of_coconuts: 0,
-                              voltage: 1.5,
-                              nailed: true };
-        assert_eq!(parrot.speed().unwrap(), 0.0);
+        parrot_test!(EuropeanParrot, tests)
     }
+
     #[test]
-    fn not_nailed_norwegian_blue_parrot() {
-        let parrot = Parrot { parrot_type: "norwegian_blue_parrot",
-                              number_of_coconuts: 0,
-                              voltage: 1.5,
-                              nailed: false };
-        assert_eq!(parrot.speed().unwrap(), 18.0);
+    fn test_african_parrot_speed() {
+        let tests = vec![
+            ParrotTest {
+                expected: 3.0,
+                config: ParrotConfig::new(1, 0.0, false),
+                description: "Test african parrot with one coconut",
+            },
+            ParrotTest {
+                expected: 0.0,
+                config: ParrotConfig::new(2, 0.0, false),
+                description: "Test african parrot with two coconuts",
+            },
+            ParrotTest {
+                expected: 12.0,
+                config: ParrotConfig::new(0, 0.0, false),
+                description: "Test african parrot with no coconuts",
+            },
+        ];
+
+        parrot_test!(AfricanParrot, tests)
```

Comment:
Per above question for `parrot_test` does this report 1 test in rust/cargo or 3 tests?

Just curious.

Reply:
See above